includes/embeds/class-amp-crowdsignal-embed-handler.php

```diff
@@ -8,59 +8,128 @@
+		// allow for the oEmbed HTML to be fetched, and can then sanitized later below.
+		$formats = [
+			'#https?://(.+\.)?polldaddy\.com/.*#i',
+			'#https?://poll\.fm/.*#i',
+			'#https?://(.+\.)?survey\.fm/.*#i', // Not available on WP 5.2.
+		];
+
+		foreach ( $formats as $format ) {
+			wp_oembed_add_provider( $format, 'https://api.crowdsignal.com/oembed', true );
+		}
 	}
 
 	/**
-	 * Unregister embed.
+	 * Unregister the embed.
+	 *
+	 * @return void
 	 */
 	public function unregister_embed() {
-		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
 	}
 
 	/**
-	 * Filter oEmbed HTML for Crowdsignal for AMP output.
+	 * Get all raw embeds from the DOM.
 	 *
-	 * @param string $cache Cache for oEmbed.
-	 * @param string $url   Embed URL.
-	 * @param array  $attr  Shortcode attributes.
-	 * @return string Embed.
+	 * @param Document $dom Document.
+	 * @return DOMNodeList|null A list of DOMElement nodes, or null if not implemented.
 	 */
-	public function filter_embed_oembed_html( $cache, $url, $attr ) {
-		$parsed_url = wp_parse_url( $url );
-		if ( empty( $parsed_url['host'] ) || empty( $parsed_url['path'] ) || ! preg_match( '#(^|\.)(?P<host>polldaddy\.com|crowdsignal\.com|survey\.fm|poll\.fm)#', $parsed_url['host'], $matches ) ) {
-			return $cache;
-		}
+	protected function get_raw_embed_nodes( Document $dom ) {
+		$queries = [
+			// For poll embeds.
+			'//iframe[ @class="cs-iframe-embed" and starts-with( @src, "https://poll.fm/" ) ]',
+			// For survey embeds.
+			'//div[ @class="pd-embed" and @data-settings ]',
+		];
 
-		$parsed_url['host'] = $matches['host'];
+		return $dom->xpath->query( implode( ' | ', $queries ) );
+	}
 
-		$output = '';
+	/**
+	 * Make embed AMP compatible.
+	 *
+	 * @param DOMElement $node DOM element.
+	 */
+	protected function sanitize_raw_embed( DOMElement $node ) {
+		$is_poll = 'cs-iframe-embed' === $node->getAttribute( 'class' );
 
-		// Poll oEmbed responses include noscript which can be used as the AMP response.
-		if ( preg_match( '#<noscript>(.+?)</noscript>#s', $cache, $matches ) ) {
-			$output = $matches[1];
+		if ( $is_poll ) {
+			$this->sanitize_poll_embed( $node );
+		} else {
+			$this->sanitize_survey_embed( $node );
 		}
+	}
 
-		if ( empty( $output ) ) {
-			if ( ! empty( $attr['title'] ) ) {
-				$name = $attr['title'];
-			} elseif ( 'survey.fm' === $parsed_url['host'] || preg_match( '#^/s/#', $parsed_url['path'] ) ) {
-				$name = __( 'View Survey', 'amp' );
-			} else {
-				$name = __( 'View Poll', 'amp' );
-			}
-			$output = sprintf( '<a href="%s" target="_blank">%s</a>', esc_url( $url ), esc_html( $name ) );
+	/**
+	 * Sanitize poll embed.
+	 *
+	 * @param DOMElement $node Poll embed.
+	 */
+	private function sanitize_poll_embed( DOMElement $node ) {
+		// Replace the `noscript` parent element with the iframe.
+		$node->parentNode->parentNode->replaceChild( $node, $node->parentNode );
```

Comment:
Should this confirm that the `$node->parentNode->parentNode` is actually a `noscript`? This could also be done done at the XPath level.